### PR TITLE
Upgrades the resolver version to lts-19.6.

### PR DESCRIPTION
**Pull request by rimbi**
Mostly applies the changes in the APIs of the dependencies.

## Purpose

Upgrade the toolchain and the dependencies to the latest version.

## Changes

Upgrades the resolver version to `lts-19.6` and ghc to `9.0.2`.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

**Comment by rimbi** [comment]
This is a work in progress but also mature enough to get your feedback.

File: haskell-src/Concordium/Crypto/ByteStringHelpers.hs [hunk 1]
```diff
@@ -113,9 +114,9 @@ instance AE.ToJSON Short65K where
 -- |JSON instances based on base16 encoding.
 instance AE.FromJSON Short65K where
   parseJSON = AE.withText "Short65K" $ \t ->
-    let (bs, rest) = BS16.decode (Text.encodeUtf8 t)
-    in if BS.null rest then return (Short65K (BSS.toShort bs))
-       else AE.typeMismatch "Not a valid Base16 encoding." (AE.String t)
+    case BS16.decode (Text.encodeUtf8 t) of
+      Right bs -> return (Short65K (BSS.toShort bs))
+      Left _ -> AE.typeMismatch "Not a valid Base16 encoding." (AE.String t)
```
Comment (abizjak) [inline haskell-src/Concordium/Crypto/ByteStringHelpers.hs:119]: For debugging it would be useful to keep the error.

File: haskell-src/Concordium/Crypto/ByteStringHelpers.hs [hunk 2]
```diff
@@ -125,9 +126,9 @@ instance AE.ToJSON ByteStringHex where
 -- |JSON instances based on base16 encoding.
 instance AE.FromJSON ByteStringHex where
   parseJSON = AE.withText "ByteStringHex" $ \t ->
-    let (bs, rest) = BS16.decode (Text.encodeUtf8 t)
-    in if BS.null rest then return (ByteStringHex (BSS.toShort bs))
-       else AE.typeMismatch "Not a valid Base16 encoding." (AE.String t)
+    case BS16.decode (Text.encodeUtf8 t) of
+      Right bs -> return (ByteStringHex (BSS.toShort bs))
+      Left _ -> AE.typeMismatch "Not a valid Base16 encoding." (AE.String t)
```
Comment (abizjak) [inline haskell-src/Concordium/Crypto/ByteStringHelpers.hs:131]: For debugging it would be useful to keep the error.

File: haskell-src/Concordium/Crypto/ByteStringHelpers.hs (hunk 1 above)
Comment (rimbi) [reply haskell-src/Concordium/Crypto/ByteStringHelpers.hs:119]: when needed it can be changed

File: haskell-src/Concordium/Crypto/ByteStringHelpers.hs (hunk 2 above)
Comment (rimbi) [reply haskell-src/Concordium/Crypto/ByteStringHelpers.hs:131]: when needed it can be changed

File: haskell-src/Concordium/Types/Transactions.hs [hunk 3]
```diff
@@ -36,6 +36,56 @@ import Concordium.Types.Updates
...
+      <> S.put thExpiry
+
+  get = do
+    thSender <- S.get
+    thNonce <- S.get
+    thEnergyAmount <- S.get
+    thPayloadSize <- S.get
+    thExpiry <- S.get
+    return $! TransactionHeader {..}
+
```
Comment (abizjak) [inline haskell-src/Concordium/Types/Transactions.hs:88]: I do not understand this change. 

This PR is meant to update the toolchains. We cannot at the same time be moving things around. That is a recipe for problems and needless conflicts.

Is this moved because of a staging restriction, or why is it necessary?

File: stack.yaml [hunk 4]
```diff
@@ -59,7 +59,8 @@ extra-lib-dirs:
 # arch: x86_64
 #
 # Extra directories used by stack for building
-# extra-include-dirs: [/path/to/dir]
+extra-include-dirs:
+    - /Library/Developer/CommandLineTools/SDKs/MacOSX12.3.sdk/usr/include/ffi/
```
Comment (abizjak) [inline stack.yaml:63]: We cannot have this in, it is a local directory on your machine.

File: haskell-src/Concordium/Types/Transactions.hs (hunk 3 above)
Comment (rimbi) [reply haskell-src/Concordium/Types/Transactions.hs:88]: I am not in favor of unnecessary changes either. Do you remember our discussion about the `deriveJSON` macro not working? It turned out to be a problem with the order of the declarations. I don't know the exact reason but with the `9.0.2` the current order of the declarations has to be changed to get rid of the compile problems.

File: stack.yaml (hunk 4 above)
Comment (rimbi) [reply stack.yaml:63]: I don't have a better solution for that atm. Do you?

File: haskell-src/Concordium/Utils.hs [hunk 5]
```diff
@@ -23,11 +23,11 @@ class (Ixed m) => At' m where
   at' :: Index m -> Lens' m (Maybe (IxValue m))
 
 instance (Hashable k, Eq k) => At' (H.HashMap k v) where
-  at' = flip H.alterF
+  at' k f = H.alterF f k
```
Comment (abizjak) [inline haskell-src/Concordium/Utils.hs:26]: Why was this necessary or useful?

Comment (rimbi) [reply haskell-src/Concordium/Utils.hs:26]: It was necessary, did not dig into the details, might be the same/similar issue to the one Thomas pointed. (eta-expansion)

Comment (td202) [inline haskell-src/Concordium/Utils.hs:26]:
```suggestion
  at' = \k f -> H.alterF f k
```
While the eta-expansion is necessary, I think expressing it as a lambda may give the optimiser more chance of inlining. Lenses in particular are often composed rather than directly applied.

**Review by td202 — APPROVED** [review]

File: haskell-src/Concordium/Utils.hs (hunk 5 above)
Comment (rimbi) [reply haskell-src/Concordium/Utils.hs:26]: Thanks for the suggestion Thomas. On the other I will not use it this time as the linter complains about it.

Comment (rimbi) [reply haskell-src/Concordium/Utils.hs:26]: And here is the error message with the original format:

```console
/Users/cemeliguzel/scratch/concordium-client/deps/concordium-base/haskell-src/Concordium/Utils.hs:26:9: error:
    • Couldn't match type: k0
                           -> H.HashMap k0 v0 -> f0 (H.HashMap k0 v0)
                     with: forall (f :: * -> *).
                           Functor f =>
                           (Maybe (IxValue (H.HashMap k v))
                            -> f (Maybe (IxValue (H.HashMap k v))))
                           -> H.HashMap k v -> f (H.HashMap k v)
      Expected: Index (H.HashMap k v)
                -> Lens' (H.HashMap k v) (Maybe (IxValue (H.HashMap k v)))
        Actual: (Maybe v0 -> f0 (Maybe v0))
                -> k0 -> H.HashMap k0 v0 -> f0 (H.HashMap k0 v0)
    • In the expression: H.alterF
      In an equation for ‘at'’: at' = H.alterF
      In the instance declaration for ‘At' (H.HashMap k v)’
    • Relevant bindings include
        at' :: Index (H.HashMap k v)
               -> Lens' (H.HashMap k v) (Maybe (IxValue (H.HashMap k v)))
          (bound at haskell-src/Concordium/Utils.hs:26:3)
   |
26 |   at' = H.alterF
   |         ^^^^^^^^
```
